### PR TITLE
Rework - change byte order access functions, LEDBuffer becomes a deque instead of a vector

### PR DESCRIPTION
**Pull request by davepl**
- Increases refresh rate to 100Hz for better display quality
- Reworks ULONGFromMemory and friends to use bswap and only where needed
- Changes the LEDBuffer collection from vector to deque
- Changes the LEDBuffer mutex from recursive to regular
- Removed IsEmpty proxy in favor of _buffer.empty to eliminate nested mutex case

### CODE REVIEW

File: globals.h [hunk 1]
```diff
@@ -14,38 +15,37 @@ constexpr auto kDefaultChainLength        = 8;
...
-            (uint64_t)payloadData[4] << 32  |
-            (uint64_t)payloadData[3] << 24  |
-            (uint64_t)payloadData[2] << 16  |
-            (uint64_t)payloadData[1] << 8   |
-            (uint64_t)payloadData[0];
+    const auto result = *(uint64_t *)payloadData;
+    if constexpr (std::endian::native == std::endian::little) 
+        return result;                      // Already little-endian, no swap needed
+    else 
+        return __builtin_bswap64(result);   // Swap for big-endian systems
```
Comment (elder-george) [inline globals.h:32]: any specific reason to not use `ntohl` and friends for this?

File: ledbuffer.h [hunk 2]
```diff
@@ -63,7 +63,7 @@ class LEDBuffer
 
   private:
 
-    std::vector<CRGB> _leds;
+    std::deque<CRGB> _leds;
```
Comment (elder-george) [inline ledbuffer.h:66]: nit: deque may be better semantically, but also requires more allocations.

if perf and memory use are critical here, vector may still be the best option, as long as you append from the end (moving will be just memmove which is usually cheap). Or you could do a ring buffer.

But most likely it won't be an issue.

File: socketserver.h [hunk 3]
```diff
@@ -110,7 +112,7 @@ class SocketServer
         _cbReceived(0)
     {
         _abOutputBuffer = std::make_unique<uint8_t []>(_maximumPacketSize);        // Must add +1 for LZ bug case in Arduino library if ever backported there
-        _pBuffer = std::make_unique<uint8_t []>(_maximumPacketSize);
+        _pBuffer        = std::make_unique<uint8_t []>(_maximumPacketSize);
         memset(&_address, 0, sizeof(_address));
```
Comment (elder-george) [inline socketserver.h:116]: nit: you could probably move all of these into the initialization block:
```C++
    SocketServer(int port, size_t maxLEDs) :
        _port(port),
        _server_fd(-1),
        _address{},
        _maximumPacketSize(STANDARD_DATA_HEADER_SIZE + LED_DATA_SIZE * maxLEDs),       
        _abOutputBuffer(std::make_unique<uint8_t []>(_maximumPacketSize),
        _pBuffer = std::make_unique<uint8_t []>(_maximumPacketSize),
        _cbReceived(0)
       {}
```
Compilers make slightly better code in such cases (and you won't need to `memset` _address below)

Note that you'll need to move `_maximumPacketSize` to be declared before both buffers for this to work

Comment (elder-george) [reply socketserver.h:116]: Also you could make some of that initialization inline
```C++
class SocketServer
{
    int                         _port;
    int                         _server_fd{-1};
    struct sockaddr_in          _address{};
    size_t _maximumPacketSize;
    std::unique_ptr<uint8_t []> _pBuffer;
    std::unique_ptr<uint8_t []> _abOutputBuffer;
public:
    size_t cbReceived{0};
public:
     SocketServer(int port, size_t maxLEDs) :
        _maximumPacketSize(STANDARD_DATA_HEADER_SIZE + LED_DATA_SIZE * maxLEDs),
       _abOutputBuffer(std::make_unique<uint8_t []>(_maximumPacketSize),
       _pBuffer = std::make_unique<uint8_t []>(_maximumPacketSize)
       {}
};
```

File: socketserver.h [hunk 4]
```diff
@@ -286,12 +291,13 @@ class SocketServer
...
             if (0 != getpeername(new_socket, (struct sockaddr *)&addr, &addr_size))
             {
+                printf("Unable to get peer name from socket!");
                 close(new_socket);
                 ResetReadBuffer();
                 continue;
             }
 
-            printf("Incoming connection from: %s", inet_ntoa(addr.sin_addr));
+            //printf("Incoming connection from: %s", inet_ntoa(addr.sin_addr));
```
Comment (elder-george) [inline socketserver.h:300]: nit: dead code

File: socketserver.h [hunk 5]
```diff
@@ -262,21 +263,25 @@ class SocketServer
...
-            int new_socket = 0;
-
             // Accept a new incoming connnection
+
+            int new_socket = 0;
             int addrlen = sizeof(_address);
             if ((new_socket = accept(_server_fd, (struct sockaddr *)&_address, (socklen_t*)&addrlen))<0)
             {
                 printf("Error accepting data!");
+                std::this_thread::sleep_for(std::chrono::milliseconds(kSocketRetryDelay));
```
Comment (elder-george) [inline socketserver.h:284]: nit: you could declare `kSocketRetryDelay` as `std::chrono::milliseconds`, then you won't need this cast here

File: socketserver.h [hunk 6]
```diff
@@ -262,21 +263,25 @@ class SocketServer
...
-                continue;
+                return false;
             }
 
-            int new_socket = 0;
-
             // Accept a new incoming connnection
+
+            int new_socket = 0;
             int addrlen = sizeof(_address);
```
Comment (elder-george) [inline socketserver.h:280]: nit: you could declare this as `socklet_t`, then you won't need cast below

**Comment by davepl** [comment]
Thanks Yury!  Very many good suggestions, I will check in an update to the PR with most of them shortly.

I’d love to see you help work on the NightDriverLED.com <http://nightdriverled.com/> code - if you’re interested, please let me know and I’ll send you Mesermizer kit (matrix, PCB, etc) to play with!

Thanks
Dave

File: globals.h (hunk 1 above)
Comment (davepl) [reply globals.h:32]: I couldn't find a 64-bit version of it, so decided to keep it orthogonal.  And I believe that when I checked the imply of ntohl it uses bswap behind the scenes.

File: socketserver.h [hunk 7]
```diff
@@ -456,13 +459,12 @@ class SocketServer
...
                         std::cerr << e.what() << '\n';
                     }
-                    
                 }
             } while (true);
 
             close(new_socket);
             ResetReadBuffer();
-            sleep(1);
+            std::this_thread::sleep_for(std::chrono::milliseconds(kSocketRetryDelay));
```
Comment (rbergen) [inline socketserver.h:467]: I don't think this cast is necessary any more, considering `kSocketRetryDelay` is now declared as an `std::chrono::millisecond` variable?

**Review by rbergen — APPROVED** [review]
I see one unnecessary cast, but otherwise LGTM!

File: ledbuffer.h [hunk 8]
```diff
@@ -127,38 +127,34 @@ class LEDBuffer
...
 public:
     explicit LEDBufferManager(size_t cBuffers)
-        : _aptrBuffers(cBuffers), _iHeadIndex(0), _iTailIndex(0), _cMaxBuffers(cBuffers)
+        : _cMaxBuffers(cBuffers)
     {}
 
     double AgeOfOldestBuffer() const
     {
-        std::lock_guard<std::recursive_mutex> lock(_mutex);  // Lock the mutex before accessing the buffer
+        std::lock_guard<std::mutex> lock(_mutex);  // Lock the mutex before accessing the buffer
```
Comment (jdoubleu) [inline ledbuffer.h:141]: You could use C++17's CTAD:
```suggestion
        std::lock_guard lock(_mutex);  // Lock the mutex before accessing the buffer
```

Comment (jdoubleu) [reply ledbuffer.h:141]: You could also use `std::scoped_lock` instead, which is a newer API feature. In this case, it wouldn't do any difference.

File: globals.h (hunk 1 above)
Comment (jdoubleu) [inline globals.h:32]: If you can use C++23, there's `std::byteswap` as a platform-independent alternative.

**Review by jdoubleu — COMMENTED** [review]
In general, I feel like the `XXXFromMemory` functions could be improved. I believe accepting a `uint8_t*` isn't the best choice. The caller needs to check for the correct size or you'll get a buffer over-read.

Since you already started using `std:;span`, it might be possible to accept a span with static extend: `std::span<const uint8_t, sizeof(DWORD)>`. Then you could also use `std::bit_cast` and `std::byteswap`. Because you factored out the type dependent functions e.g. `__builtin_bswap32` with a `std::byteswap`, all these functions can be combined into a single template.

---
Instead of using `uint8_t` to represent bytes, you could use `std::byte` instead, which is a dedicated type rather than a type alias. However, that may cause some headache down the line with C APIs. Ultimately it's personal preference.

**Comment by davepl** [comment]
Thought I caught that!  Thanks!

**Pull request merged**